### DISCUSSION TITLE
Callbacks at exiting

### DISCUSSION THREAD

**Discussion by rickyyx** [shadow/shadow#429] [General]
Hi, would love to print out some debug information from my application at the end of the simulation. 

When using libevent,  and code after the evbase_dispatch() is never run.  And I have seen from the code and commits that there seems to be way to register callbacks at the end of the simulation. Any pointers for such? 

Thanks!

**Comment by robgjansen** [comment]
Shadow internally keeps a queue of functions that should get called when your application ends (which would normally happen at the end of the simulation). Shadow then iterates through the list of such functions and calls them sequentially:
https://github.com/shadow/shadow/blob/1190a345acd8260aa197e6b10c0498db588ea6ac/src/main/host/shd-process.c#L883-L904

To add callbacks to the list of those that get called, you can use the standard `on_exit()` or `atexit()` sys calls.

`on_exit`:
https://github.com/shadow/shadow/blob/1190a345acd8260aa197e6b10c0498db588ea6ac/src/main/host/shd-process.c#L4710

`atexit`:
https://github.com/shadow/shadow/blob/1190a345acd8260aa197e6b10c0498db588ea6ac/src/main/host/shd-process.c#L4722

Please keep in mind that these functions have probably not been tested very well, so there may be bugs.

**Comment by rickyyx** [comment]
Alright! This is really helpful. Thanks!
